src/dbnode/storage/index/aggregate_results.go

```diff
@@ -0,0 +1,237 @@
+// THE SOFTWARE.
+
+package index
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/pool"
+)
+
+const missingDocumentFields = "invalid document fields: empty %s"
+
+type aggregatedResults struct {
+	sync.RWMutex
+
+	nsID          ident.ID
+	aggregateOpts AggregateResultsOptions
+
+	resultsMap *AggregateResultsMap
+
+	idPool    ident.Pool
+	bytesPool pool.CheckedBytesPool
+
+	pool       AggregateResultsPool
+	valuesPool AggregateValuesPool
+}
+
+// NewAggregateResults returns a new AggregateResults object.
+func NewAggregateResults(
+	namespaceID ident.ID,
+	aggregateOpts AggregateResultsOptions,
+	opts Options,
+) AggregateResults {
+	return &aggregatedResults{
+		nsID:          namespaceID,
+		aggregateOpts: aggregateOpts,
+		resultsMap:    newAggregateResultsMap(opts.IdentifierPool()),
+		idPool:        opts.IdentifierPool(),
+		bytesPool:     opts.CheckedBytesPool(),
+		pool:          opts.AggregateResultsPool(),
+		valuesPool:    opts.AggregateValuesPool(),
+	}
+}
+
+func (r *aggregatedResults) Reset(
+	nsID ident.ID,
+	aggregateOpts AggregateResultsOptions,
+) {
+	r.Lock()
+
+	r.aggregateOpts = aggregateOpts
+
+	// finalize existing held nsID
+	if r.nsID != nil {
+		r.nsID.Finalize()
+	}
+
+	// make an independent copy of the new nsID
+	if nsID != nil {
+		nsID = r.idPool.Clone(nsID)
+	}
+	r.nsID = nsID
+
+	// reset all values from map first
+	for _, entry := range r.resultsMap.Iter() {
+		valueMap := entry.Value()
+		valueMap.reset()
+	}
+
+	// reset all keys in the map next
+	r.resultsMap.Reset()
+
+	// NB: could do keys+value in one step but I'm trying to avoid
+	// using an internal method of a code-gen'd type.
+	r.Unlock()
+}
+
+func (r *aggregatedResults) AddDocuments(batch []doc.Document) (int, error) {
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size := r.resultsMap.Len()
+	r.Unlock()
+	return size, err
+}
+
+func (r *aggregatedResults) addDocumentsBatchWithLock(
+	batch []doc.Document,
+) error {
+	for i := range batch {
+		err := r.addDocumentWithLock(batch[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addDocumentWithLock(
+	document doc.Document,
+) error {
+	for _, field := range document.Fields {
+		if err := r.addFieldWithLock(field.Name, field.Value); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addFieldWithLock(
+	term []byte,
+	value []byte,
+) error {
+	if len(term) == 0 {
+		return fmt.Errorf(missingDocumentFields, "term")
+	}
+
+	if len(value) == 0 {
+		return fmt.Errorf(missingDocumentFields, "value")
+	}
+
+	// NB: can cast the []byte -> ident.ID to avoid an alloc
+	// before we're sure we need it.
+	var termID ident.ID = ident.BytesID(term)
+
+	// if a term filter is provided, ensure this field matches the filter,
+	// otherwise ignore it.
+	if len(r.aggregateOpts.TermFilter) > 0 {
```

Comment:
nit: mind making this a function on the type? something like 

```
type AggregateResultsOptions  struct {
  ...
  TermFilter [][]byte
}

type AggregateTermFilter [][]byte

func (a AggregateTermFilter) Filter(term []byte) bool {
  if len(a) == 0 {
    return false 
  }
  for ...
  return true
}
```

Reply:
SGTM, guess then we can use it elsewhere :)